File: ts/packages/knowPro/src/textLocationIndex.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { ListIndexingResult, TextLocation } from "./interfaces.js";
import { IndexingEventHandlers } from "./interfaces.js";
import {
    addTextBatchToEmbeddingIndex,
    addTextToEmbeddingIndex,
    EmbeddingIndex,
    indexOfNearestTextInIndex,
    indexOfNearestTextInIndexSubset,
    TextEmbeddingIndexSettings,
} from "./fuzzyIndex.js";
import { NormalizedEmbedding } from "typeagent";
import { Scored } from "./common.js";

export type ScoredTextLocation = {
    score: number;
    textLocation: TextLocation;
};

export interface ITextToTextLocationIndex {
    addTextLocation(
        text: string,
        textLocation: TextLocation,
    ): Promise<ListIndexingResult>;
    addTextLocations(
        textAndLocations: [string, TextLocation][],
        eventHandler?: IndexingEventHandlers,
    ): Promise<ListIndexingResult>;
    lookupText(
        text: string,
        maxMatches?: number,
        thresholdScore?: number,
    ): Promise<ScoredTextLocation[]>;

    serialize(): ITextToTextLocationIndexData;
    deserialize(data: ITextToTextLocationIndexData): void;
}

export interface ITextToTextLocationIndexData {
    textLocations: TextLocation[];
    embeddings: Float32Array[];
}

export class TextToTextLocationIndex implements ITextToTextLocationIndex {
    private textLocations: TextLocation[];
    private embeddingIndex: EmbeddingIndex;

    constructor(public settings: TextEmbeddingIndexSettings) {
        this.textLocations = [];
        this.embeddingIndex = new EmbeddingIndex();
    }

    public get size(): number {
        return this.embeddingIndex.size;
    }

    public get(pos: number): TextLocation {
        return this.textLocations[pos];
    }

    public async addTextLocation(
        text: string,
        textLocation: TextLocation,
    ): Promise<ListIndexingResult> {
        const result = await addTextToEmbeddingIndex(
            this.embeddingIndex,
            this.settings.embeddingModel,
            [text],
        );
        if (result.numberCompleted > 0) {
            this.textLocations.push(textLocation);
        }
        return result;
    }

    public async addTextLocations(
        textAndLocations: [string, TextLocation][],
        eventHandler?: IndexingEventHandlers,
        batchSize?: number,
    ): Promise<ListIndexingResult> {
        const indexingEvents = createMessageIndexingEventHandler(
            textAndLocations,
            eventHandler,
        );
        const result = await addTextBatchToEmbeddingIndex(
            this.embeddingIndex,
            this.settings.embeddingModel,
            textAndLocations.map((tl) => tl[0]),
            batchSize ?? this.settings.batchSize,
            indexingEvents,
        );
        if (result.numberCompleted > 0) {
            textAndLocations =
                result.numberCompleted === textAndLocations.length
                    ? textAndLocations
                    : textAndLocations.slice(0, result.numberCompleted);
            this.textLocations.push(...textAndLocations.map((tl) => tl[1]));
        }
        return result;
    }

    public async lookupText(
        text: string,
        maxMatches?: number,
        thresholdScore?: number,
    ): Promise<ScoredTextLocation[]> {
        const matches = await indexOfNearestTextInIndex(
            this.embeddingIndex,
            this.settings.embeddingModel,
            text,
            maxMatches,
            thresholdScore,
        );
        return matches.map((m) => {
            return {
                textLocation: this.textLocations[m.item],
                score: m.score,
            };
        });
    }

    /**
     * Find text locations nearest to the provided text.
     * But only search over the subset of text locations in this index identified by ordinalsToSearch
     * @param text
     * @param ordinalsToSearch
     * @param maxMatches
     * @param thresholdScore
     * @returns
     */
    public async lookupTextInSubset(
        text: string,
        ordinalsToSearch: number[],
        maxMatches?: number,
        thresholdScore?: number,
    ): Promise<ScoredTextLocation[]> {
        const matches = await indexOfNearestTextInIndexSubset(
            this.embeddingIndex,
            this.settings.embeddingModel,
            text,
            ordinalsToSearch,
            maxMatches,
            thresholdScore,
        );
        return matches.map((m) => {
            return {
                textLocation: this.textLocations[m.item],
                score: m.score,
            };
        });
    }

    public lookupByEmbedding(
        textEmbedding: NormalizedEmbedding,
        maxMatches?: number,
        thresholdScore?: number,
        predicate?: (messageOrdinal: number) => boolean,
    ): ScoredTextLocation[] {
        const matches = this.embeddingIndex.getIndexesOfNearest(
            textEmbedding,
            maxMatches,
            thresholdScore,
            predicate,
        );
        return this.toScoredLocations(matches);
    }

    public lookupInSubsetByEmbedding(
        textEmbedding: NormalizedEmbedding,
        ordinalsToSearch: number[],
        maxMatches?: number,
        thresholdScore?: number,
    ): ScoredTextLocation[] {
        const matches = this.embeddingIndex.getIndexesOfNearestInSubset(
            textEmbedding,
            ordinalsToSearch,
            maxMatches,
            thresholdScore,
        );
        return this.toScoredLocations(matches);
    }

    public clear(): void {
        this.textLocations = [];
        this.embeddingIndex.clear();
    }

    public serialize(): ITextToTextLocationIndexData {
        return {
            textLocations: this.textLocations,
            embeddings: this.embeddingIndex.serialize(),
        };
    }

    public deserialize(data: ITextToTextLocationIndexData): void {
        if (data.textLocations.length !== data.embeddings.length) {
            throw new Error(
                `TextToTextLocationIndexData corrupt. textLocation.length ${data.textLocations.length} != ${data.embeddings.length}`,
            );
        }
        this.textLocations = data.textLocations;
        this.embeddingIndex.deserialize(data.embeddings);
    }

    private toScoredLocations(matches: Scored[]): ScoredTextLocation[] {
        return matches.map((m) => {
            return {
                textLocation: this.textLocations[m.item],
                score: m.score,
            };
        });
    }
}

function createMessageIndexingEventHandler(
    textAndLocations: [string, TextLocation][],
    eventHandler?: IndexingEventHandlers | undefined,
): IndexingEventHandlers | undefined {
    return eventHandler && eventHandler.onTextIndexed !== undefined
        ? {
              onEmbeddingsCreated: (texts, batch, batchStartAt) => {
                  eventHandler.onTextIndexed!(
                      textAndLocations,
                      textAndLocations.slice(
                          batchStartAt,
                          batchStartAt + batch.length,
                      ),
                      batchStartAt,
                  );
                  return true;
              },
          }
        : eventHandler;
}

File: python/ta/typeagent/knowpro/textlocationindex.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from dataclasses import dataclass
from typing import Any, Protocol

from ..aitools.embeddings import NormalizedEmbeddings
from ..aitools.vectorbase import VectorBase
from .importing import TextEmbeddingIndexSettings
from .interfaces import IndexingEventHandlers, ListIndexingResult, TextLocation


@dataclass
class ScoredTextLocation:
    text_location: TextLocation
    score: float


class ITextToTextLocationIndexData(Protocol):
    textLocations: list[TextLocation]
    embeddings: NormalizedEmbeddings


class ITextToTextLocationIndex(Protocol):
    async def add_text_location(
        self, text: str, text_location: TextLocation
    ) -> ListIndexingResult:
        raise NotImplementedError

    async def add_text_locations(
        self,
        text_and_locations: list[tuple[str, TextLocation]],
        event_handler: IndexingEventHandlers | None = None,
    ) -> ListIndexingResult:
        raise NotImplementedError

    async def lookup_text(
        self,
        text: str,
        max_matches: int | None = None,
        threshold_score: float | None = None,
    ) -> list[ScoredTextLocation]:
        raise NotImplementedError

    def serialize(self) -> ITextToTextLocationIndexData:
        raise NotImplementedError

    def deserialize(self, data: ITextToTextLocationIndexData) -> None:
        raise NotImplementedError


class TextToTextLocationIndex(ITextToTextLocationIndex):
    def __init__(self, settings: TextEmbeddingIndexSettings):
        self._text_locations: list[TextLocation] = []
        self._vector_base: VectorBase = VectorBase()

    def __len__(self) -> int:
        return len(self._vector_base)

    def __bool__(self) -> bool:
        return True

    def get(self, pos: int, default: TextLocation | None = None) -> TextLocation | None:
        size = len(self._text_locations)
        if -size <= pos < size:
            return self._text_locations[pos]
        return default

    async def add_text_location(
        self, text: str, text_location: TextLocation
    ) -> ListIndexingResult:
        # TODO: Catch errors
        await self._vector_base.add_key(text)
        self._text_locations.append(text_location)
        return ListIndexingResult(1)

    async def add_text_locations(
        self,
        text_and_locations: list[tuple[str, TextLocation]],
        event_handler: IndexingEventHandlers | None = None,
        batch_size: int | None = None,
    ) -> ListIndexingResult:
        # TODO: Honor batch size
        # TODO: Catch errors
        await self._vector_base.add_keys([text for text, _ in text_and_locations])
        self._text_locations.extend([location for _, location in text_and_locations])
        return ListIndexingResult(len(text_and_locations))

    async def lookup_text(
        self,
        text: str,
        max_matches: int | None = None,
        threshold_score: float | None = None,
    ) -> list[ScoredTextLocation]:
        matches = await self._vector_base.fuzzy_lookup(
            text, max_hits=max_matches, min_score=threshold_score
        )
        return [
            ScoredTextLocation(self._text_locations[match.ordinal], match.score)
            for match in matches
        ]

    async def lookup_text_in_subset(
        self,
        *args,
        **kwds,
    ) -> Any:
        raise NotImplementedError

    def serialize(self) -> Any:
        raise NotImplementedError  # TODO implement TextToTextLocationIndex serialization

    def deserialize(self, data: Any) -> None:
        raise NotImplementedError  # TODO: implement TextToTextLocationIndex deserialization
